libqrexec/exec.c

```diff
@@ -36,66 +41,89 @@ void register_exec_func(do_exec_t *func) {
 void exec_qubes_rpc_if_requested(char *prog, char *const envp[]) {
     /* avoid calling qubes-rpc-multiplexer through shell */
     if (strncmp(prog, RPC_REQUEST_COMMAND, RPC_REQUEST_COMMAND_LEN) == 0) {
-        char *tok;
+        char *tok, *saveptr;
         char *argv[16]; // right now 6 are used, but allow future extensions
         size_t i = 0;
-
-        tok=strtok(prog, " ");
+        tok=strtok_r(prog, " ", &saveptr);
         do {
             if (i >= sizeof(argv)/sizeof(argv[0])-1) {
                 fprintf(stderr, "To many arguments to %s\n", RPC_REQUEST_COMMAND);
                 exit(1);
             }
             argv[i++] = tok;
-        } while ((tok=strtok(NULL, " ")));
+        } while ((tok=strtok_r(NULL, " ", &saveptr)));
         argv[i] = NULL;
         argv[0] = QUBES_RPC_MULTIPLEXER_PATH;
         execve(QUBES_RPC_MULTIPLEXER_PATH, argv, envp);
         perror("exec qubes-rpc-multiplexer");
-        exit(1);
+        _exit(126);
     }
 }
 
 void fix_fds(int fdin, int fdout, int fderr)
 {
     int i;
-    for (i = 0; i < 256; i++)
+    for (i = 3; i < 256; i++)
         if (i != fdin && i != fdout && i != fderr)
             close(i);
-    dup2(fdin, 0);
-    dup2(fdout, 1);
-    dup2(fderr, 2);
-    close(fdin);
-    close(fdout);
-    if (fderr != 2)
-        close(fderr);
+    if (dup2(fdin, 0) || dup2(fdout, 1) || dup2(fderr, 2) ||
+        close(fdin) || close(fdout) || (fderr != 2 && close(fderr)))
+        abort();
 }
 
-void do_fork_exec(const char *cmdline, int *pid, int *stdin_fd, int *stdout_fd,
-        int *stderr_fd)
+int do_fork_exec(char *cmdline,
```

Comment:
I like this improvement in error reporting, but you don't really check the return value anywhere.

Reply:
Also, it may be worth reporting other errors this way too, instead of `exit()`.